### ISSUE TITLE
Support for "stacked" connectors

### ISSUE THREAD

**Issue by gwr69** [modelica/ModelicaSpecification#3206]
## Motivation

Regarding the support for connecting components with a multitude of `connector` classes my understanding as modeler and library developer of a system dynamics library (i.e., [BusinessSimulation](https://github.com/bslMS/BusinessSimulation)) is the following:

1. `Composite connectors` are hierarchical connectors intended for conveniently connecting many sub-connectors _at once_. Sometimes the need for making individual connections for elements within a `composite connector` arise (see #2543)
2. `Expandable connectors` are typically _flexible_ connectors with elements being dynamically added during the modeling process, which make building large models much more convenient. While they are excellent for building larger models, I am a bit hesitant in using them for pre-built components, e.g., library classes.

The restrictions and caveats associated with the above mentioned mechanisms and the way connections are typically established in dedicated SD tools (with naturally more simple use cases) has me wonder about a simpler possibility, which I would like to describe below.

## Stacked Connectors (while `block connectors` are not implemented yet)

Let's look at a very simple structure in Modelica (System Modeler was used in this case):

<img src="https://user-images.githubusercontent.com/17766670/171626231-ffe64558-5330-48ed-b3e5-1bef3ebe1c6a.png" width="250"></img>

```
model ModelConnections
  model component1
    Modelica.Blocks.Interfaces.RealOutput y1;
    Modelica.Blocks.Interfaces.RealOutput y2;
  end component1;

  component1 c1;

  model component2
    Modelica.Blocks.Interfaces.RealInput u1;
    Modelica.Blocks.Interfaces.RealInput u2;
  end component2;

  component2 c2;
equation
  connect(c1.y2, c2.u2);
end ModelConnections;
```

It is perfectly legal and thus simply a matter of (likely _improper_) _style_ to "stack" connectors _atop of each other_ in diagrams and icons. In the given example, Wolfram SystemModeler's GUI (I have not checked this for other environments) will not know _which_ connector is selected in DiagramView and so it is probably just the order of the connectors being instantiated in the model, which determines the actual instances being connected in the connection equation (e.g., `c1.y2` is connected `c2.u2` with the GUI's Connect Line Tool when the stacked connectors are selected).

**Wouldn't it be possible to allow tools to recognize the _proximity_ of graphical representations as to provide a choice dialog for selecting the appropriate connector?** 

This would not have any other implications as the model's code will not be affected at all—it is merely a convenient and flexible way to prevent diagram clutter and "porcupine-like" icons. In _inductive_ modeling approaches like system dynamics such multiplicity of connectors arises quickly (mainly, but not exclusively `causal` ones) and most often these connectors will not be used "_en bloc_".

Maybe this could be specified more formally by something like an `annotation`, e.g., To maybe make this more formal, one may discuss an `annotation` like `representativeConnector` (or even more general `representativeInstance` ?) which tells a tool to not show a separate graphical representation in the icon or diagram, but to instead use the graphical representation of the declared representative instance as _proxy_ for _establishing a selection in the tool's front end_.

Essentially, this would make the use of many individual connectors (not meant to be treated as a single block) more convenient.

**Comment by gwr69** [comment]
This issue is not receiving a lot of attention, so let me ask a couple of questions to either stimulate discussion or come to closure more quickly:

1. Is there anything that makes this proposal _less preferable_ than the `connector block` (working title according to #2543 )?

2. Having a less restrictive `connector block` as an alternative to a `composite connector` has been discussed last in 2020. Is there a reason for something in this vein not being followed up on?

3. If the answer to 2 is "no", what may be a realistic timeline for one of these solutions being added in the specs?

**Comment by gwr69** [comment]
Thank you for the reply and I can quite see the lower priority for making more elaborate changes in the vein of `connector block`. But you have rather nicely confirmed my hunch, that tools may _already_ support "stacked/overlapping" connectors in establishing connections as this is totally within specification. There has been "hesitance" out of fear this may break something between tools, which can now be alleviated by reference to Dymola. ;-)

**Comment by HansOlsson** [comment]
As for "block connector" I would see the following:

- They can directly only contain connectors - not flow (or across) variables.
- You can connect to the whole or parts (since it is intentional we don't have the usual problem of breaking encapsulation). Note that we might also have a similar easing of the restriction for expandable connectors.

In this way "connector block" would be similar to "expandable connector" but with fixed contents.

**Comment by gwr69** [comment]
> As for "block connector" I would see the following:
> 
> * They can directly only contain connectors - not flow (or across) variables.
> * You can connect to the whole or parts (since it is intentional we don't have the usual problem of breaking encapsulation). Note that we might also have a similar easing of the restriction for expandable connectors.
> 
> In this way "connector block" would be similar to "expandable connector" but with fixed contents.

Weren‘t acausal connectors allowed within `expandable connector` ? (d‘accord otherwise)

**Comment by HansOlsson** [comment]
> > As for "block connector" I would see the following:
> > 
> > * They can directly only contain connectors - not flow (or across) variables.
> > * You can connect to the whole or parts (since it is intentional we don't have the usual problem of breaking encapsulation). Note that we might also have a similar easing of the restriction for expandable connectors.
> > 
> > In this way "connector block" would be similar to "expandable connector" but with fixed contents.
> 
> Weren‘t acausal connectors allowed within `expandable connector` ? (d‘accord otherwise)

Yes, they are allowed but as sub-connectors - so I don't see whether that would be different.

**Comment by gwr69** [comment]
Ok, there are a couple of thinks to ponder for clarity (at least for my head):

- Would the prefix `connector` be needed and accordingly then `input` or `output` prefixes? (similarity to `expandable connector` suggests that this were not the case)
- Could the internal connectors (e.g., variables) be potentially present only? (as with expandable ones)
- Would the `block connector` just be a mere notional “collector”, i.e., so connections to `m.bc.c` are legal, which would then be dissimilar; my understanding is that `m.bc.c` would really be `m.c` so `bc` can be completely eliminated from a parsed model)
